### PR TITLE
Add Jest tests, schema rework, various other things

### PR DESCRIPTION
**Pull request by tsa96**
This PR gets a ton of things in the rewrite up to a standard where it's no longer a complete headfuck to work with.

* I've gone through the whole Prisma schema by hand fixing up all the relationships and removing autogenerated shit. The structure should be the same as the old API now.
* I've improved all our DTO handling. Things now use [class-validator](https://github.com/typestack/class-validator) and [class-transformer](https://github.com/typestack/class-transformer) for data validation and transformation.
* I've set Jest up and starting porting tests.
* Various other small things

I'm not going to go into everything in too much detail here, it should have all been discussed in meetings and Discord already.

Please be hyper-critical of all the changes here, I want us to establish standard practices for everything that will last. Once this is in I'll feel a lot happier about bringing on more people to work on this.

### CODE REVIEW

File: server/src/main.ts [hunk 1]
```diff
@@ -24,10 +25,11 @@ async function bootstrap() {
 
     const config = new DocumentBuilder()
         .setTitle('Momentum Mod API')
-        .setDescription('The Momentum Mod API - Made with 💖')
+        .setDescription('The Momentum Mod API - Made with 🦵')
```
Comment (Gocnak) [inline server/src/main.ts:28]: Made with leg emoji?

File: server/src/@common/dto/user/user.dto.ts [hunk 2]
```diff
@@ -1,52 +1,94 @@
...
     updatedAt: Date;
 
+    private _avatar: string;
+
+    @ApiProperty()
     get avatarURL(): string {
-        const bans = this.bans;
-        const isAvatarBanned = bans & (1 << 2); // TODO: Refactor however needed to use Ban 'enum' (cyclic dep issue occurs when requiring user model in this file)
-        if (isAvatarBanned) {
+        // TODO: apparently steam's hosting has changed and this stuff is on cloudfare now, ask goc
```
Comment (Gocnak) [inline server/src/@common/dto/user/user.dto.ts:47]: It has. The base URL is now:
`https://avatars.cloudflare.steamstatic.com/b0d98c45eb07061557a0ca85a7c3fb7cd24fda71_full.jpg`
where the old was:
`https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/b0/b0d98c45eb07061557a0ca85a7c3fb7cd24fda71_full.jpg`

So we should only be storing the `b0d98c45eb07061557a0ca85a7c3fb7cd24fda71_full.jpg` part now.

File: server/src/@common/dto/user/user.dto.ts [hunk 3]
```diff
@@ -1,52 +1,94 @@
...
+            return this._avatar
+                ? `https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/${this.alias}`
+                : null;
         }
     }
+
     set avatarURL(val: string) {
-        const newVal = val?.replace('https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/', '');
-        this._avatar = newVal;
+        this._avatar = val?.replace('https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/', '');
```
Comment (Gocnak) [inline server/src/@common/dto/user/user.dto.ts:58]: `val?.replace('https://avatars.cloudflare.steamstatic.com/', '');`

File: server/src/@common/dto/user/user.dto.ts [hunk 4]
```diff
@@ -1,52 +1,94 @@
...
-        if (isAvatarBanned) {
+        // TODO: apparently steam's hosting has changed and this stuff is on cloudfare now, ask goc
+        if (this.bans & EBan.BANNED_AVATAR) {
             return appConfig.baseURL + '/assets/images/blank_avatar.jpg';
         } else {
-            const avatar = this._avatar;
-            return avatar ? `https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/${avatar}` : null;
+            return this._avatar
+                ? `https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/${this.alias}`
+                : null;
```
Comment (Gocnak) [inline server/src/@common/dto/user/user.dto.ts:53]: `https://avatars.cloudflare.steamstatic.com/${this._avatar}` (and not `this.alias`)

File: .gitignore [hunk 5]
```diff
@@ -16,6 +16,7 @@ mysql/
 
 # dependencies
 /node_modules
+/package-lock.json
```
Comment (Gocnak) [inline .gitignore:19]: Yes, it should. CI (dependabot, etc) needs it. The official node documentation say to include it in every repo: https://nodejs.dev/learn/the-package-lock-json-file

**Review by Gocnak — CHANGES_REQUESTED** [review]

File: server/src/main.ts (hunk 1 above)
Comment (tsa96) [reply server/src/main.ts:28]: Fuck, I was hoping this wouldn't be noticed :( 
I would rather we remove this than have it at all, it's meaningless and corny

File: server/src/@common/dto/user/user.dto.ts (hunk 2 above)
Comment (tsa96) [reply server/src/@common/dto/user/user.dto.ts:47]: Oh sorry, yeah I actually fixed this yesterday just hadn't include in this branch. Made some changes to the DTOs yesterday that I think are better but I wanted to work with them some more before committing, the URL change is in there.

Comment (tsa96) [reply server/src/@common/dto/user/user.dto.ts]: Okay I've pushed the changed DTOs I was doing including the URL change

File: server/src/@common/dto/map/map.dto.ts [hunk 6]
```diff
@@ -1,46 +1,72 @@
...
+        console.log(JSON.stringify(_map));
 
         let submitter = _submitter;
         if (submitter == null) {
             // if null then try get it from map object
-            submitter = (_map as any).users;
+            submitter = (_map as any).user;
         }
-        Logger.log(JSON.stringify(submitter));
+        console.log(JSON.stringify(submitter));
```
Comment (Gocnak) [inline server/src/@common/dto/map/map.dto.ts:62]: Console logs instead of `Logger.log`?

File: server/src/main.ts (hunk 1 above)
Comment (tsa96) [reply server/src/main.ts]: Nuked it

File: .gitignore (hunk 5 above)
Comment (tsa96) [reply .gitignore]: Apologies yeah, I was reading up on this earlier and reading just that (i.e. that you should include it), dunno why I still included the commit

File: server/src/@common/dto/map/map.dto.ts (hunk 6 above)
Comment (TheNerdMan) [reply server/src/@common/dto/map/map.dto.ts:62]: Yeah we should be using the `Logger` class, not the `console.log()`. Unless it was for dev and these should have been removed before the PR

File: server/src/modules/users/users.service.ts [hunk 7]
```diff
@@ -1,55 +1,74 @@
...
+        skip?: number,
+        take?: number,
+        expand?: string[],
+        search?: string,
+        playerID?: string,
+        playerIDs?: string[]
+    ): Promise<PagedResponseDto<UserDto[]>> {
+        const where: Prisma.UserWhereInput = {};
+
+        if (playerID) {
```
Comment (TheNerdMan) [inline server/src/modules/users/users.service.ts:36]: Potential bug here, what happens when you provide a playerID and playerIDs? Depending on the order by in the repo you may not always get the same response. I'd swap these around, too
```ts
if (playerIDs) {
    where.steamID = { in: playerIDs };
} else if (playerID) {
   take = 1;
    where.steamID = playerID;
}
```
As then if you somehow provide both the more specific request is always honoured (as its quicker). Or just throw a bad request error if a client provides both

File: server/src/@common/dto/map/map.dto.ts (hunk 6 above)
Comment (tsa96) [reply server/src/@common/dto/map/map.dto.ts:62]: Logger.log wasn't actually working, it's just debug anyway, this is an unfinished thing Alex was working on I wanted to just get cleaned up. I really wouldn't worry about debug stuff like this for now

**Review by TheNerdMan — CHANGES_REQUESTED** [review]

File: server/src/modules/users/users.service.ts (hunk 7 above)
Comment (tsa96) [reply server/src/modules/users/users.service.ts:36]: Dunno why I can't respond to the logger comment, but that logging stuff is on your code, it's already on the main branch. It's just random debug, I wanted to use it and Logger.log doesn't seem to work with Jest currently. I assume it'll get deleted in the future by you, I wouldn't worry about it

Comment (tsa96) [reply server/src/modules/users/users.service.ts:36]: Yeah, should just throw a bad request. Made it do that in last push

**Comment by tsa96** [comment]
Just done a big update to do the DTOs a way Alex and me are both happy with, glad to get that sorted. Also addressed all the above stuff, should be in a good state now.

**Review by TheNerdMan — CHANGES_REQUESTED** [review]
~~So I've just pulled this down, run `npm i` and `npx prisma generate` to update my `@prisma/client` to have the correct classes and it seems that `e2e-environments.ts` and `seed.ts` are broken from the schema changes.~~

Once these are sorted I think this PR is ready to be merged.

Scratch that, a node_module clean (delete and regenerate) fixes this. PR done from my perspective

**Review by TheNerdMan — APPROVED** [review]

**Review by Gocnak — APPROVED** [review]
Nits I noted in voice, once those are pushed then 👍

File: server/src/modules/users/users.repo.ts [hunk 8]
```diff
@@ -21,28 +21,36 @@ export class UsersRepo {
...
         const count = await this.prisma.user.count({
             where: where
         });
+
         const users = await this.prisma.user.findMany({
             where: where,
-            skip: skip != null ? +skip : undefined,
-            take: take != null ? +take : undefined
+            include: include,
+            skip: skip,
```
Comment (TheNerdMan) [inline server/src/modules/users/users.repo.ts:37]: This with its removed `+` (e.g `skip: +skip`) still still causes a Prisma runtime error

Comment (TheNerdMan) [reply server/src/modules/users/users.repo.ts:37]: Example error:
```
Invalid `this.prisma.user.findMany()` invocation in
F:\Users\alexd\Repos\GitHub\Others\momentum-mod\website-nestJS\server\src\modules\users\users.repo.ts:34:46

  31     where: where
  32 });
  33
→ 34 const users = await this.prisma.user.findMany({
       where: {},
       include: {
         profile: undefined,
         userStats: undefined
       },
       skip: '0',
             ~~~
       take: '20'
             ~~~~
     })

Argument skip: Got invalid value '0' on prisma.findManyUser. Provided String, expected Int.
Argument take: Got invalid value '20' on prisma.findManyUser. Provided String, expected Int.
```

**Review by TheNerdMan — CHANGES_REQUESTED** [review]

File: server/src/modules/users/users.repo.ts (hunk 8 above)
Comment (tsa96) [reply server/src/modules/users/users.repo.ts:37]: Huh, is this coming from a test or running manually? Never seen this myself, ParseIntPipe should be handling it

Comment (TheNerdMan) [reply server/src/modules/users/users.repo.ts:37]: I ran this manually but tests would have caught it too, I just didn't use them yet

**Review by TheNerdMan — APPROVED** [review]

**Pull request merged**